src/Api/PubnubApi/EventEngine/EventEngine.cs

```diff
@@ -0,0 +1,749 @@
+					LogCallback?.Invoke($"Received Message ({index + 1} of {receiveMessages.Length}) : {JsonConvert.SerializeObject(receiveMessages[index])}");
+					if (receiveMessages[index].Channel.IndexOf("-pnpres") > 0)
+					{
+						var presenceData = JsonConvert.DeserializeObject<PresenceEvent>(receiveMessages[index].Payload.ToString());
+					}
+					else
+					{
+						LogCallback?.Invoke($"Message : {JsonConvert.SerializeObject(receiveMessages[index].Payload)}");
+						PNMessageResult<object> messageResult = new PNMessageResult<object>();
+						messageResult.Channel = receiveMessages[index].Channel;
+						messageResult.Message = receiveMessages[index].Payload;
+						AnnounceMessage?.Invoke(messageResult);
+					}
+				}
+			}
+		}
+	}
+
+	public class HandshakeFailed : EffectInvocation
+	{
+		public List<string> Channels { get; set; }
+		public List<string> ChannelGroups { get; set; }
+	}
+	public class CancelHandshakeFailed : EffectInvocation
+	{
+
+	}
+    #endregion
+    public enum EventType
+	{
+		SubscriptionChanged,
+		SubscriptionRestored,
+		Handshake,
+		CancelHandshake,
+		HandshakeSuccess,
+		ReceiveMessages,
+		CancelReceiveMessages,
+		ReceiveSuccess,
+		HandshakeFailure,
+		CancelHandshakeFailure,
+		ReceiveFailure,
+		ReceiveReconnect,
+		CancelReceiveReconnect,
+		ReceiveReconnectSuccess,
+		ReceiveReconnectFailure,
+		ReceiveReconnectGiveUp,
+		HandshakeReconnect,
+		CancelHandshakeReconnect,
+		HandshakeReconnectSuccess,
+		HandshakeReconnectFailure,
+		HandshakeReconnectGiveUp,
+		ReconnectionFailed,
+		Disconnect,
+		Reconnect
+	}
+
+	public class ExtendedState
+	{
+		public List<string> Channels { get; set; }
+		public List<string> ChannelGroups { get; set; }
+		public long? Timetoken { get; set; }
+		public int? Region { get; set; }
+		public int Attempts { get; set; }
+
+		public ExtendedState()
+		{
+			Channels = new List<string>();
+			ChannelGroups = new List<string>();
+			Timetoken = 0;
+			Region = 0;
+			Attempts = 0;
+		}
+
+	}
+
+	public class EventEngine
+	{
+		private EventEngine pnEventEngine = null;
+
+		public ExtendedState Context;
+		public State? CurrentState { get; set; }
+		public List<State> States { get; set; }
+
+		public EffectDispatcher Dispatcher;
+
+		public EventEmitter Emitter;
+		public IPubnubUnitTest PubnubUnitTest { get; set; }
+
+		public EventEngine(EffectDispatcher dispatcher, EventEmitter emitter)
+		{
+			if (PubnubUnitTest != null )
+			{
+				PubnubUnitTest.EventTypeList?.Clear();
+			}
+			this.Dispatcher = dispatcher;
+			States = new List<State>();
+			Context = new ExtendedState();
+			this.Emitter = emitter;
+			emitter.RegisterHandler(this.Transition);
+		}
+
+		public State CreateState(StateType type)
+		{
+			var newState = new State(type);
+			if (States.Find(s=> s.StateType == type) != null) 
+			{
+				throw new InvalidOperationException($"StateType = {type} already exist.");
+			}
+			States.Add(newState);
+			return newState;
+		}
+
+		public void Transition(Event e)
```

Comment:
So It seems like we need to start with here, 
Because as per our updated state transition description, The function may follow below signature
`transition :: (State s, Event ev) => (State z, Invocation[] fs)`

And so this will lead to tweak other path - to follow Invocations and invocation structures instead `Event` that we have here!